contracts/modules/staking/StakingProducts.sol

```diff
@@ -0,0 +1,460 @@
+  uint public constant MAX_ACTIVE_TRANCHES = 8; // 7 whole quarters + 1 partial quarter
+  uint public constant WEIGHT_DENOMINATOR = 100;
+
+  // pool id => product id => Product
+  mapping(uint => mapping(uint => IStakingPool.StakedProduct)) private _products;
+  // pool id => { effectiveWeight, targetWeight }
+  mapping(uint => Weights) public weights;
+
+  address public immutable coverContract;
+  address public immutable stakingPoolFactory;
+
+  constructor(address _coverContract, address _stakingPoolFactory) {
+    coverContract = _coverContract;
+    stakingPoolFactory = _stakingPoolFactory;
+  }
+
+  function onlyManager(address stakingPool) internal view {
+    address manager = IStakingPool(stakingPool).manager();
+    if (msg.sender != manager) {
+      revert IStakingPool.OnlyManager();
+    }
+  }
+
+  function onlyStakingPool(address stakingPool) private view {
+    if (msg.sender != stakingPool) {
+      revert OnlyStakingPool();
+    }
+  }
+
+  function products(uint poolId, uint productId)
+  external
+  view
+  override
+  returns (IStakingPool.StakedProduct memory) {
+    return _products[poolId][productId];
+  }
+
+  function getStakingPoolAddress(uint poolId) private view returns (address pool) {
+    pool = address(ICover(coverContract).stakingPool(poolId));
+    // TODO: is this always valid?
+    if (pool.code.length == 0) {
+      revert InvalidStakingPool();
+    }
+  }
+
+  function recalculateEffectiveWeights(uint poolId, uint[] calldata productIds) external {
+    address stakingPool = getStakingPoolAddress(poolId);
+    (
+    uint globalCapacityRatio,
+    /* globalMinPriceRatio */,
+    /* initialPriceRatios */,
+    /* capacityReductionRatios */
+    uint[] memory capacityReductionRatios
+    ) = ICover(coverContract).getPriceAndCapacityRatios(productIds);
+
+    uint _totalEffectiveWeight = weights[poolId].totalEffectiveWeight;
+
+    for (uint i = 0; i < productIds.length; i++) {
+      uint productId = productIds[i];
+      IStakingPool.StakedProduct memory _product = _products[poolId][productId];
+
+      uint16 previousEffectiveWeight = _product.lastEffectiveWeight;
+      _product.lastEffectiveWeight = getEffectiveWeight(
+        IStakingPool(stakingPool),
+        productId,
+        _product.targetWeight,
+        globalCapacityRatio,
+        capacityReductionRatios[i]
+      );
+      _totalEffectiveWeight = _totalEffectiveWeight - previousEffectiveWeight + _product.lastEffectiveWeight;
+      _products[poolId][productId] = _product;
+    }
+    weights[poolId].totalEffectiveWeight = _totalEffectiveWeight.toUint32();
+  }
+
+  function setProducts(uint poolId, StakedProductParam[] memory params) external {
+    address stakingPool = getStakingPoolAddress(poolId);
+    onlyManager(stakingPool);
+
+    uint globalCapacityRatio;
+    uint globalMinPriceRatio;
+    uint[] memory initialPriceRatios;
+    uint[] memory capacityReductionRatios;
+
+    {
+      uint numProducts = params.length;
```

Comment:
ran into stack-too-deep errors in this function....it's slightly modified from the original.